expr/tagquery/expression.go

```diff
@@ -45,134 +40,346 @@ func (e Expressions) Strings() []string {
+	RequiresNonEmptyValue() bool
+
+	// Matches takes a string which should either be a tag key or value depending on the return
+	// value of OperatesOnTag(), then it returns whether the given string satisfies this expression
+	Matches(string) bool
+
+	// MatchesExactly returns a bool to indicate whether the key / value of this expression (depending
+	// on OperatesOnTag()) needs to be an exact match with the key / value of the metrics it evaluates
+	// F.e:
+	// in the case of the expression "tag1=value1" we're only looking for metrics where the value
+	// associated with tag key "tag1" is exactly "value1", so a simple string comparison is sufficient.
+	// in other cases like "tag1=~val.*" or "tag^=val" this isn't the case, a simple string comparison
+	// is not sufficient to decide whether a metric should be part of the result set or not.
+	// since simple string comparisons are cheaper than other comparison methods, whenever possible we
+	// want to use string comparison.
+	MatchesExactly() bool
+
+	// GetMetricDefinitionFilter returns a MetricDefinitionFilter
+	// The MetricDefinitionFilter takes a metric definition, looks at its tags and returns a decision
+	// regarding this query expression applied to its tags
+	GetMetricDefinitionFilter(lookup IdTagLookup) MetricDefinitionFilter
+
+	// StringIntoBuilder takes a builder and writes a string representation of this expression into it
+	StringIntoBuilder(builder *strings.Builder)
+}
+
+// ParseExpression returns an expression that's been generated from the given
+// string, in case of an error the error gets returned as the second value
+func ParseExpression(expr string) (Expression, error) {
+	var pos int
+	prefix, regex, not := false, false, false
+	resCommon := expressionCommon{}
+
+	// scan up to operator to get key
+FIND_OPERATOR:
+	for ; pos < len(expr); pos++ {
+		switch expr[pos] {
+		case '=':
+			break FIND_OPERATOR
+		case '!':
+			not = true
+			break FIND_OPERATOR
+		case '^':
+			prefix = true
+			break FIND_OPERATOR
+		case ';':
+			return nil, InvalidExpressionError(expr)
+		}
 	}
 
-	res.RequiresNonEmptyValue = true
-	if expression[equalPos-1] == '!' {
-		operatorStartPos = equalPos - 1
-		res.RequiresNonEmptyValue = false
-		res.Operator = NOT_EQUAL
-	} else if expression[equalPos-1] == '^' {
-		operatorStartPos = equalPos - 1
-		res.Operator = PREFIX
-	} else {
-		operatorStartPos = equalPos
-		res.Operator = EQUAL
+	// key must not be empty
+	if pos == 0 {
+		return nil, InvalidExpressionError(expr)
 	}
 
-	res.Key = expression[:operatorStartPos]
-	err := validateQueryExpressionTagKey(res.Key)
+	resCommon.key = expr[:pos]
+	err := validateQueryExpressionTagKey(resCommon.key)
 	if err != nil {
-		return res, fmt.Errorf("Error when validating key \"%s\" of expression \"%s\": %s", res.Key, expression, err)
+		return nil, fmt.Errorf("Error when validating key \"%s\" of expression \"%s\": %s", resCommon.key, expr, err)
```

Comment:
We should a prefix to this and all other errors contained inside `expr` and `tagquery` like we do with the indexes, such as `expr-tagquery: ` or `tagquery: `. This will allow us to more easily filter logs when looking for problems.

Reply:
I was under the impression that we're only prefixing log statements with prefixes such as `expr-tagquery:`, because this helps in understanding the logs. Are we doing the same for returned error values which get returned to the user?

Reply:
Good point, no I don't think we are.